lib/renderer/extensions/i18n.ts

```diff
@@ -1,59 +1,58 @@
-'use strict'
-
 // Implementation of chrome.i18n.getMessage
 // https://developer.chrome.com/extensions/i18n#method-getMessage
 //
 // Does not implement predefined messages:
 // https://developer.chrome.com/extensions/i18n#overview-predefined
 
-const ipcRendererUtils = require('@electron/internal/renderer/ipc-renderer-internal-utils')
+import * as ipcRendererUtils from '@electron/internal/renderer/ipc-renderer-internal-utils'
+
+interface Placeholder {
+  content: string;
+  example?: string;
+}
 
-const getMessages = (extensionId) => {
+const getMessages = (extensionId: number) => {
   try {
     const data = ipcRendererUtils.invokeSync('CHROME_GET_MESSAGES', extensionId)
-    return JSON.parse(data) || {}
+    return JSON.parse(data as any) || {}
   } catch {
     return {}
   }
 }
 
-const replaceNumberedSubstitutions = (message, substitutions) => {
+const replaceNumberedSubstitutions = (message: string, substitutions: string[]) => {
   return message.replace(/\$(\d+)/, (_, number) => {
     const index = parseInt(number, 10) - 1
     return substitutions[index] || ''
   })
 }
 
-const replacePlaceholders = (message, placeholders, substitutions) => {
-  if (typeof substitutions === 'string') {
-    substitutions = [substitutions]
-  }
-  if (!Array.isArray(substitutions)) {
-    substitutions = []
-  }
+const replacePlaceholders = (message: string, placeholders: Record<string, Placeholder>, substitutions: string[] | string) => {
+  if (typeof substitutions === 'string') substitutions = [substitutions]
+  if (!Array.isArray(substitutions)) substitutions = []
 
   if (placeholders) {
-    Object.keys(placeholders).forEach((name) => {
+    Object.keys(placeholders).forEach((name: string) => {
       let { content } = placeholders[name]
-      content = replaceNumberedSubstitutions(content, substitutions)
+      content = replaceNumberedSubstitutions(content, substitutions as string[])
```

Comment:
This `as string[]` shouldn't be required, it should be known at this point that `substitutions` is definitely an array.🤔 In order to prove it you may need to make a `const` redef

```js
const substitutionsArray = Array.isArray(substitutions) ? substitutions : []
```

Which is const and of type `string[]`